python/ray/rllib/models/modelv2.py

```diff
@@ -0,0 +1,153 @@
+        options (dict): options for the model, documented in ModelCatalog
+    """
+
+    def __init__(self, obs_space, action_space, num_outputs, options):
+        """Initialize the model.
+
+        This method should create any variables used by the model.
+        """
+
+        self.obs_space = obs_space
+        self.action_space = action_space
+        self.num_outputs = num_outputs
+        self.options = options
+
+    def get_initial_state(self):
+        """Get the initial recurrent state values for the model.
+
+        Returns:
+            list of np.array objects, if any
+        """
+        return []
+
+    def forward(self, input_dict, state, seq_lens):
+        """Call the model with the given input tensors and state.
+
+        Any complex observations (dicts, tuples, etc.) will be unpacked by
+        __call__ before being passed to forward(). To access the flattened
+        observation tensor, refer to input_dict["obs_flat"].
+
+        This method can be called any number of times. In eager frameworks,
+        each call to forward() will eagerly evaluate the model. In symbolic
+        frameworks, each call to forward creates a computation graph that
+        operates over the variables of this model (i.e., shares weights).
+
+        Custom models should override this instead of __call__.
+
+        Arguments:
+            input_dict (dict): dictionary of input tensors, including "obs",
+                "obs_flat", "prev_action", "prev_reward", "is_training"
+            state (list): list of state tensors with sizes matching those
+                returned by get_initial_state + the batch dimension
+            seq_lens (list): 1d tensor holding input sequence lengths
+
+        Returns:
+            (outputs, state, feature_layer): The model output tensor of size
+                [BATCH, num_outputs], a list of state tensors of sizes
+                [BATCH, state_size_i], and a tensor of [BATCH, feature_size]
+        """
+        raise NotImplementedError
+
+    def get_branch_output(self, branch_type, num_outputs, feature_layer=None):
+        """Get the branch output of the model (e.g., "value" branch).
+
+        It is important to note that the method outputs are tied to the
+        immediately previous call to forward(). This means that after calling
+        forward() once, you must retrieve all its branch outputs before calling
+        forward() again.
+
+        Arguments:
+            branch_type (str): identifier for the branch (e.g., "value")
+            num_outputs (int): size of the branch output
+            feature_layer (tensor): if specified, this hints that the branch
+                output should be computed using this shared feature layer.
+                However custom models are free to ignore this hint.
+
+        Returns:
+            tensor: branch output of size [BATCH, num_outputs]
+        """
+        raise NotImplementedError
+
+    def custom_loss(self, policy_loss, loss_inputs):
```

Comment:
Assuming this model class is supposed to be more generic, I think this should not take in a `policy_loss` argument.

Reply:
Seems like this model could serve as a more abstract model class, e.g. to be used for value-functions. In that case, for example, I wouldn't expect it to take in a `policy_loss` as an argument since it doesn't necessarily need to know anything about the policy loss.

Or maybe I misunderstand what the model really is or how it's supposed to be used.

Reply:
The assumption here is that each algorithm only ever has one model, which encompasses any possible sub models such as value functions internally as branch outputs. This avoids the problem of some algorithms needing multiple models, without loss of generality via custom models.

Reply:
Okay, in that case the `policy_loss` makes sense.

Are we sure that the one model model is what we want or is it possible that this becomes a limitation at some point (soon)? I can easily imagine situations where I might want to decouple the models from each other. Imagine for example that I train an actor-critic model and only want to store the actor from it (and discard the critic). I guess it's possible to dig the policy from the encapsulating model, but feels a bit clumsy.

At least I think there should be a plan to allow separating the models in the future if this indeed becomes a problem.

Reply:
Yeah, probably we can solve that issue with variable naming. Currently get weights returns a flat array, but it could instead return a dict that can be filtered to include only the relevant variables.

Reply:
Okay yeah, I think that sounds reasonable.